BUG: Compiling C extension against numpy 2.0.1 fails with `error: 'I' undeclared`

### Describe the issue:

Attempting to build the `gstlal-inspiral` package (C modules, C extension modules, and pure Python modules built using autotools) fails in compiling a C extension module, in an environment with numpy 2.0.1 with:

```text
/home/conda/feedstock_root/build_artifacts/gstlal-inspiral_1722339059186/work/python/_spawaveform.c: In function 'SPAWaveformReduceSpin':
/home/conda/feedstock_root/build_artifacts/gstlal-inspiral_1722339059186/work/python/_spawaveform.c:585:55: error: 'I' undeclared (first use in this function)
  585 |         hOfF[k] = amp * (cos(Psi+shft*f+phi0+piBy4) - I*sin(Psi+shft*f+phi0+piBy4));
      |                                                    
```

The `I` in question is supposed to be that provided by `complex.h`.

This was briefly discussed [here](https://github.com/numpy/numpy/pull/26789#issuecomment-2258531119) (and replies).

### Reproduce the code example:

```python
n/a
```

The source code for the extension module in question is [here](https://git.ligo.org/lscsoft/gstlal/-/blob/gstlal-inspiral-1.10.0-v1/gstlal-inspiral/python/_spawaveform.c?ref_type=tags#L1-L25).

### Error message:

```shell
/home/conda/feedstock_root/build_artifacts/gstlal-inspiral_1722339059186/work/python/_spawaveform.c: In function 'SPAWaveformReduceSpin':
/home/conda/feedstock_root/build_artifacts/gstlal-inspiral_1722339059186/work/python/_spawaveform.c:585:55: error: 'I' undeclared (first use in this function)
  585 |         hOfF[k] = amp * (cos(Psi+shft*f+phi0+piBy4) - I*sin(Psi+shft*f+phi0+piBy4));
      |
```


### Python and NumPy Versions:

Python 3.10.4
Numpy 2.0.1

### Runtime Environment:

_No response_

### Context for the issue:

This briefly prevented forward motion on https://github.com/conda-forge/gstlal-inspiral-feedstock/pull/24 (until the addition of a [patch](https://github.com/conda-forge/gstlal-inspiral-feedstock/pull/24/commits/247dffebc610acf3ad697b36fed7be432cc7b998)) (that build is eventually foiled by other issues, but the fundamental problem here stands independently, I think).

Numba is also affected by this issue. Numba 0.60.0 fails to build from source:
```shell
In file included from numba/_helpermod.c:23:
  numba/_helperlib.c: In function ‘numba_cpowf’:
  numba/_helperlib.c:22:44: error: ‘I’ undeclared (first use in this function)
     22 |     #define _complex_float_ctor(r, i) (r + I * i)
        |                                            ^
  numba/_helperlib.c:147:12: note: in expansion of macro ‘_complex_float_ctor’
    147 |     *out = _complex_float_ctor((float) _out.real, (float) _out.imag);
        |            ^~~~~~~~~~~~~~~~~~~
  numba/_helperlib.c:22:44: note: each undeclared identifier is reported only once for each function it appears in
     22 |     #define _complex_float_ctor(r, i) (r + I * i)
        |                                            ^
  numba/_helperlib.c:147:12: note: in expansion of macro ‘_complex_float_ctor’
    147 |     *out = _complex_float_ctor((float) _out.real, (float) _out.imag);
        |            ^~~~~~~~~~~~~~~~~~~
  error: command '/usr/bin/gcc' failed with exit code 1
```

--- a/doc/source/numpy_2_0_migration_guide.rst
+++ b/doc/source/numpy_2_0_migration_guide.rst
@@ -220,6 +220,19 @@ using the NumPy types. You can still write cython code using the ``c.real`` and
 ``c.imag`` attributes (using the native typedefs), but you can no longer use
 in-place operators ``c.imag += 1`` in Cython's c++ mode.
 
+Because NumPy 2 now includes ``complex.h`` code that uses a variable named
+``I`` may see an error such as
+
+.. code-block::C
+   error: expected ‘)’ before ‘__extension__’
+                    double I,
+
+to use the name ``I`` requires an ``#undef I`` now.
+
+.. note::
+  NumPy 2.0.1 briefly included the ``#undef I`` to help users not already
+  including ``complex.h``.
+
 
 Changes to namespaces
 =====================

--- a/doc/source/reference/c-api/types-and-structures.rst
+++ b/doc/source/reference/c-api/types-and-structures.rst
@@ -1611,3 +1611,29 @@ for completeness and assistance in understanding the code.
    ``arrayobject.h`` header. This type is not exposed to Python and
    could be replaced with a C-structure. As a Python type it takes
    advantage of reference- counted memory management.
+
+
+NumPy C-API and C complex
+=========================
+When you use the NumPy C-API, you will have access to complex real declarations
+``npy_cdouble`` and ``npy_cfloat``, which are declared in terms of the C
+standard types from ``complex.h``. Unfortunately, ``complex.h`` contains
+`#define I ...`` (where the actual definition depends on the compiler), which
+means that any downstream user that does ``#include <numpy/arrayobject.h>``
+could get ``I`` defined, and using something like declaring ``double I;`` in
+their code will result in an obscure compiler error like
+
+.. code-block::C
+    error: expected ‘)’ before ‘__extension__’
+                    double I,
+
+This error can be avoided  by adding::
+
+    #undef I
+
+to your code.
+
+.. versionchanged:: 2.0
+    The inclusion of ``complex.h`` was new in NumPy 2, so that code defining
+    a different ``I`` may not have required the ``#undef I`` on older versions.
+    NumPy 2.0.1 briefly included the ``#under I``

--- a/numpy/_core/include/numpy/npy_common.h
+++ b/numpy/_core/include/numpy/npy_common.h
@@ -379,11 +379,6 @@ typedef struct
 
 #include <complex.h>
 
-// Downstream libraries like sympy would like to use I
-// see https://github.com/numpy/numpy/issues/26787
-#ifdef I
-#undef I
-#endif
 
 #if defined(_MSC_VER) && !defined(__INTEL_COMPILER)
 typedef _Dcomplex npy_cdouble;

